### PR TITLE
Cubeb: rewrite locking

### LINKED ISSUE #12528: [Regression] RPCS3 crashes if airpods are taken off on macOS.
### Quick summary

After #12246 if I take my airpods off while a game is running, RPCS3 crashes. 

### Details

After #12246 if I take my airpods off while a game is running, RPCS3 crashes. This does not occur when I change my default audio device before taking them off, or with wired headphones. Disconnecting by turning off bluetooth also does not trigger the issue.

### Build with regression

v0.0.23-13872-98b730c8

### Attach two log files

#12246 (bad) : [RPCS3.log](https://github.com/RPCS3/rpcs3/files/9378081/RPCS3.log)
 #12300 (good) : [RPCS3.log](https://github.com/RPCS3/rpcs3/files/9378083/RPCS3.log)

### System configuration

macOS 13.0 beta (22A5295i)
Apple M1 chip
Airpods Pro

### PR DESCRIPTION
**Pull request by Vestrel**
Fixes #12528 and maybe #12798

### CODE REVIEW

File: rpcs3/Emu/Audio/Cubeb/CubebBackend.cpp [hunk 1]
```diff
@@ -80,8 +81,18 @@ bool CubebBackend::Operational()
 
 bool CubebBackend::DefaultDeviceChanged()
 {
-	std::lock_guard lock{m_dev_sw_mutex};
-	return !m_reset_req.observe() && m_default_dev_changed;
+	if (m_default_device.empty() || m_reset_req.observe()) {
```
Comment (Megamouse) [inline rpcs3/Emu/Audio/Cubeb/CubebBackend.cpp:84]: Newline

File: rpcs3/Emu/Cell/Modules/cellAudio.cpp [hunk 2]
```diff
@@ -191,6 +192,13 @@ u32 audio_ringbuffer::backend_write_callback(u32 size, void *buf)
 	return static_cast<u32>(cb_ringbuf.pop(buf, size, true));
 }
 
+void audio_ringbuffer::backend_state_callback(AudioStateEvent event)
+{
+	if (event == AudioStateEvent::DEFAULT_DEVICE_MAYBE_CHANGED) {
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellAudio.cpp:197]: Newline

File: rpcs3/Emu/Audio/Cubeb/CubebBackend.cpp [hunk 3]
```diff
@@ -427,14 +432,14 @@ long CubebBackend::data_cb(cubeb_stream* stream, void* user_ptr, void const* /*
 
 	std::unique_lock lock(cubeb->m_cb_mutex, std::defer_lock);
 
-	if (stream != cubeb->m_stream)
+	if (!cubeb->m_reset_req.observe() && lock.try_lock_for(std::chrono::microseconds{50}) && cubeb->m_write_callback && cubeb->m_playing)
```
Comment (elad335) [inline rpcs3/Emu/Audio/Cubeb/CubebBackend.cpp:435]: Chancing a lock for a period of time needs to be well justified otherwise it looks like aworaround.

Comment (elad335) [reply rpcs3/Emu/Audio/Cubeb/CubebBackend.cpp:435]: And unlike the standard library, shared_mutex has the method is_free() which can test the ability to obtain a lock. You can write such a test before the if and call yield a single time if it's currently owned.

Comment (Vestrel) [reply rpcs3/Emu/Audio/Cubeb/CubebBackend.cpp:436]: It's typically taken by nonblocking code for a short period of time, so it makes sense to spin a bit.

**Pull request merged**